neo/Consensus/Helper.cs

```diff
@@ -26,8 +27,16 @@ internal static class Helper
         public static bool CommitSent(this IConsensusContext context) => context.CommitPayloads[context.MyIndex] != null;
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static bool BlockSent(this IConsensusContext context) => context.Block != null;
+
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public static bool ViewChanging(this IConsensusContext context) => context.ChangeViewPayloads[context.MyIndex]?.GetDeserializedMessage<ChangeView>().NewViewNumber > context.ViewNumber;
+        public static bool ViewChanging(this IConsensusContext context) => context.ChangeViewPayloads[context.MyIndex]?.GetDeserializedMessage<ChangeView>().NewViewNumber > context.ViewNumber && !context.MoreThanFNodesCommitted();
+
+        // A possible attack can happen if the last node to commit is malicious and either sends change view after his
+        // commit to stall nodes in a higher view, or if he refuses to send recovery messages. In addition, if a node
+        // asking change views loses network or crashes and comes back when nodes are committed in more than one higher
+        // numbered view, it is possible for the node accepting recovery and commit in any of the higher views, thus
+        // potentially splitting nodes among views and stalling the network.
+        public static bool MoreThanFNodesCommitted(this IConsensusContext context) => context.CommitPayloads.Count(p => p != null) > context.F();
```

Comment:
This can result in a deadlock because there may have been more than `F` nodes committed, but the local node did not receive the messages.

Reply:
If more than `F` nodes committed they periodically send `RecoveryMessage` it will not cause deadlock. Also we have tested it and it does not deadlock.

Reply:
This code has been tested and finally we got a better experience after using this PR. No deadlocks still detected.

Reply:
There is a problem though, this needs to count commits in all views. We likely need a change to fix that.

Reply:
Thanks to @vncoelho for pointing out that the way it is counting commits won't work properly (and was only working in the 4 node scenario)

Reply:
You can't assume that the message is deliverable.

Reply:
Can’t assume that which message is deliverable? The recovery message that is repeatedly sent every block time from a committed node? Or what message are you referring to?

Reply:
Code changes are needed to ensure that commits are counted properly, but with some adjustments it will be able to count commits in all views properly.

Reply:
Code adjustments have been made. Resolving this one to wait for comments now on the latest version.